tests/pytests/unit/returners/test_prometheus_textfile_return.py

```diff
@@ -0,0 +1,126 @@
+        "fun_args": ["applyme"],
+        "success": True,
+    }
+
+    expected = "\n".join(
+        sorted(
+            [
+                "# HELP salt_procs Number of salt minion processes running",
+                "# TYPE salt_procs gauge",
+                "salt_procs 0",
+                "# HELP salt_states_succeeded Number of successful states in the run",
+                "# TYPE salt_states_succeeded gauge",
+                "salt_states_succeeded 2",
+                "# HELP salt_states_failed Number of failed states in the run",
+                "# TYPE salt_states_failed gauge",
+                "salt_states_failed 0",
+                "# HELP salt_states_changed Number of changed states in the run",
+                "# TYPE salt_states_changed gauge",
+                "salt_states_changed 2",
+                "# HELP salt_states_total Total states in the run",
+                "# TYPE salt_states_total gauge",
+                "salt_states_total 2",
+                "# HELP salt_states_success_pct Percent of successful states in the run",
+                "# TYPE salt_states_success_pct gauge",
+                "salt_states_success_pct 100.0",
+                "# HELP salt_states_failure_pct Percent of failed states in the run",
+                "# TYPE salt_states_failure_pct gauge",
+                "salt_states_failure_pct 0.0",
+                "# HELP salt_states_changed_pct Percent of changed states in the run",
+                "# TYPE salt_states_changed_pct gauge",
+                "salt_states_changed_pct 100.0",
+                "# HELP salt_elapsed_time Time spent for all operations during the state run",
+                "# TYPE salt_elapsed_time gauge",
+                "salt_elapsed_time 13.695",
+                "# HELP salt_last_started Estimated time the state run started",
+                "# TYPE salt_last_started gauge",
+                "# HELP salt_last_completed Time of last state run completion",
+                "# TYPE salt_last_completed gauge",
+                "",
+            ]
+        )
+    )
+
+    opts = temp_salt_minion.config.copy()
+    opts["cachedir"] = cache_dir
+    with patch("salt.returners.prometheus_textfile.__opts__", opts, create=True), patch(
+        "salt.returners.prometheus_textfile.__salt__", {}, create=True
+    ):
+        prometheus_textfile.returner(ret)
+
+        with salt.utils.files.fopen(
+            os.path.join(cache_dir, "prometheus_textfile", "salt.prom")
+        ) as prom_file:
+            salt_prom = prom_file.read()
+
+        # Drop time-based fields for comparison
+        salt_prom = "\n".join(
+            sorted(
+                [
+                    line
+                    for line in salt_prom.split("\n")
+                    if not line.startswith("salt_last_started")
+                    and not line.startswith("salt_last_completed")
+                ]
+            )
+        )
+        assert salt_prom == expected
```

Comment:
**suggestion (non-blockng)**: I *would* like to see a test for 'add_state_name`, as well as for changing the filename (note that both existing and proposed changes for filename handling with names like `foo.bar.baz` treat only `baz` as the extension (that is, `something.tar.gz` has an extension of `.gz`, not `.tar.gz`.)

It's not a blocker for me as I've tested that locally, but from a long-term perspective, having those tests would help avoid introducing inadvertent bugs.

Reply:
What's the best way to start on that? Seems like it might be an integration test. There aren't any good examples of returner testing in the pytests directory.

Reply:
If you'd like, I can go over this on Thursday morning's [Test Clinic](https://www.twitch.tv/saltprojectoss/schedule) (even if you can't make it live, there will be a recording). I think a unit, or at most a functional test would be fine here... for `add_state_name` it would basically be the same as the test that you have, just with the different expected output. For the filename part, something like `foo.bar.baz` and ensure that the filename becomes `foo.bar-whatever.baz`. It would be pretty easy to stick in a parametrization like so:

```
@pytest.mark.parametrize('statename, filename, expected_filename', [
    ('whatever', 'foo.bar.baz', 'foo.bar-whatever.baz'),
    ('cool', 'foo.bar.baz', 'foo.bar-cool.baz'),
    ...  # other kinds go here - with extensions, without, with just "one" extension...
])
def test_when_add_statename_is_added_filename_should_be_correctly_altered(statename, filename, expected_filename):
    ...
    assert actual_filename == expected_filename
```

which would get a couple of different examples of what we want to see.

Reply:
@nicholasmhughes after starting to give an example of what I was talking about, I was 99% there so I just went ahead and finished writing the tests that I was talking about - I opened a PR against your branch in your fork :+1:

Reply:
:heart_eyes: